src/basic/fs-util.c

```diff
@@ -1066,6 +1066,11 @@ int access_fd(int fd, int mode) {
         return r;
 }
 
+void unlink_tempfilep(char (*p)[]) {
+        if (!endswith(*p, ".XXXXXX"))
+                unlink(*p);
```

Comment:
maybe cast the unlink() call to void? also, I think it would make a lot of sense to use unlink_noerrno() here, so that the cleanup doesn't change errno

Reply:
also I think this could use a comment, because it's not clear to the uninitiated what this is supposed to do

Reply:
`(void)` yes, but `unlink_noerrno` no. None of our functions are supposed to preserve `errno`.

Reply:
generally our functions should not care, but the ones specifically for the purpose of usage within `_cleanup_()` should probably leave errno unmodified, because their effect is mostly hidden after all...

I mean, consider this piece of code:

```c
{
        for (;;) {
                …
                if (some_syscall() < 0)
                        break;
                …
        }
}
if (errno == ENOENT)
        return 0;
return -errno;
```

The loop above does something until some error is seen on some syscall, and that error is then process. All good with the above... But now if the loop is changed to include `_cleanup_(unlink_tempfilep) char something[] = …` at the top, then suddenly everything breaks, in a very non-obvious way, because the unlink() is run when the `break` is seen, possibly overriding the errno that is expected to be used afterwards. issues like this are really hard to see...

I figure this might be something we should fix for other such clean-up functions too, however at least for rmdir_and_freep() and unlink_and_freep() we already do this properly, and we should really do this here, too

Reply:
→ #9038.